Widgets: Adds support for Simple Payment Buttons as Widgets

rodrigoi opened the pull request:
**DO NOT MERGE - THIS A FEATURE BRANCH**

Depends on:

- Simple Payments: adds plan support check for Jetpack and WordPress.com (#9650 and D14005-code)

#### Changes proposed in this Pull Request:

Adds basic support on the Customizer to add Simple Payment Buttons as Sidebar Widgets.

| Widget List | Widget Customizer Form & Preview |
| - | - |
| ![screen shot 2018-05-18 at 12 14 37](https://user-images.githubusercontent.com/233601/40242922-5af9c0ac-5a95-11e8-9929-945e2c4e7e21.png) | ![screen shot 2018-05-18 at 12 15 34](https://user-images.githubusercontent.com/233601/40242925-5fb172b6-5a95-11e8-889b-4b5763f08170.png) |

#### Testing instructions:

Pre-requisites:
- apply this branch on a Jetpack site with an active Business Plan subscription and a Theme with at least one widget area
- to create one or more Payment Buttons on the Post/Page editor

then you'll need to:

- open the Customizer on your sandboxed Business Site
- go to Widgets and select a Widget Area
- click on Add Widget
- search on the widget panel for Simple Payment
- select the Simple Payment widget

The Simple Payment Customizer should list the available Payment Buttons, and the Customizer Preview should show the item selected by default. All changes should update live on the preview window and the site should only be updated upon publishing.

Known Issues

- sometimes the live update removes the PayPal button.
- if there are no Payment Buttons, it just doesn't show a drop down.
- if the selected Payment Button is deleted, the widget just shows the title and empty content.

#### Proposed changelog entry for your changes:

Commit:
Widgets: Adds support for Simple Payment Buttons as Widgets

Copons commented:
@rodrigoi This tests well!
One nitpick/personal preference about code organization: I find it a bit confusing that we have the main file (`simple-payments.php` outside the `/simple-payments` folder.

Though, apparently there's just no established pattern across our widgets.
Some have all files in their folders, some have all files in the root, and others are like this widget.

Since we're introducing a new widget from scratch, I'd rather listen what @Automattic/jetpack folks have to say on this one, in order to ship the most up-to-date-and-awesome code possible.

Also, since I've pinged them, I'd like to ask what are we supposed to do to bring this files into JP Fusion.
Is it enough to [whitelist](https://github.com/Automattic/jetpack/blob/master/fusion/whitelist.json) them and call it a day?
In this case let's remember to do so now or maybe when we're closer to release.

kwight commented on modules/widgets/simple-payments/form.php:
```diff
@@ -0,0 +1,21 @@
+		<?php } ?>
+	</select>
+</p>
+<?php } ?>
```
This sort of thing feels quite fragile to me; closing a bracket that's been presumably opened elsewhere outside of the current file. Is there a reason for splitting this form out? I'd consider more "standard" widget coding to keep this all in the `widget()` method, or place it in another method of the `Jetpack_Simple_Payments_Widget` class and pass it $product_posts` (although not splitting it out still seems the best option to me).

Copons replied:
Might be wrong, but this bracket looks like is closing `<?php if ( ! empty( $product_posts ) ) { ?>` at line 10 of the same file?

jeherve commented on modules/widgets/simple-payments.php:
```diff
@@ -0,0 +1,107 @@
+				'customize_selective_refresh' => true,
+			);
+
+			parent::__construct( 'Jetpack_Simple_Payments_Widget', __( 'Simple Payments', 'jetpack' ), $widget );
```
Could you add the `` filter here?

rodrigoi replied:
Updated on 5d749a37cc859351c0ee28fb728683d1cdd9d930.

jeherve commented on modules/widgets/simple-payments.php:
```diff
@@ -0,0 +1,107 @@
+		function widget( $args, $instance ) {
+			echo $args['before_widget'];
+
+			$title = apply_filters( 'widget_title', $instance['title'] );
```
Could you `/** This filter is documented in core/src/wp-includes/default-widgets.php */` here, to make our parser happy?

rodrigoi replied:
Added on 5d749a37cc859351c0ee28fb728683d1cdd9d930. Also added the `jetpack_widget_name` filter to the constructor with it's comment.

jeherve commented on modules/widgets/simple-payments.php:
```diff
@@ -0,0 +1,107 @@
+			if( ! empty( $instance['product_post_id'] ) ) {
+				$attrs = array( 'id' => $instance['product_post_id'] );
+			} else {
+				$product_posts = get_posts( [
```
This `[]` notation is not going to be okay in Jetpack, since we support PHP 5.2+.

rodrigoi replied:
Updated on 0b74235429d74625242a9ea780fd5077e4feb711.

jeherve commented on modules/widgets/simple-payments.php:
```diff
@@ -0,0 +1,107 @@
+		 * @param array $instance Previously saved values from database.
+		 */
+		function form( $instance ) {
+			$product_posts = get_posts( [
```
Same as above.

rodrigoi replied:
Uprated on 0b74235429d74625242a9ea780fd5077e4feb711.

jeherve commented on modules/widgets/simple-payments.php:
```diff
@@ -0,0 +1,107 @@
+	
+	// Register Jetpack_Simple_Payments_Widget widget.
+	function register_widget_jetpack_simple_payments() {
+		register_widget( 'Jetpack_Simple_Payments_Widget' );
```
Should we maybe not register the widget at all when Simple Payments are not available on the site?

rodrigoi replied:
Any suggestion on how to proceed? `simple-payment` does not appear to be on the plans features, so `active_plan_supports` is not an option. Also tried `shortcode_exists`, but widgets are initialized before modules.
Added Jetpack activation check on 3461e3569cf7a7994f46afdde87d7f5a847a851a.

jeherve reviewed (dismissed):
A few drive-by comments, I hope this helps a bit.

kwight commented:
I get lots of warnings and errors; maybe this is because the widget is getting registered when Simple Payments itself is not active? This site has only a personal plan (which I don't believe includes SP):

![screen shot 2018-05-18 at 9 30 35 am](https://user-images.githubusercontent.com/349751/40247347-edf8e460-5a80-11e8-9d27-37e85856864e.png)

> One nitpick/personal preference about code organization: I find it a bit confusing that we have the main file simple-payments.php outside the /simple-payments folder.

Because the widget should only be registered if the SP module is active, I'd agree that it should all be under `/modules/simple-payments`. `simple-payments-widget.php` can have the widget class, and it can be registered from `init_hook_action` (similar to how the shortcode is being handled). I've also left another comment about not splitting out the form.

Commit:
Widgets: adds stats tracking for the Simple Payments Widget view

Commit:
Widgets: Add filters and actions documentation warnings

Commit:
Widgets: changed array syntax from [] to array()

Commit:
Widgets: removes magic strings from the Simple Payment Widget, and add jetpack activation check

Commit:
Widgets: adds default properties to prevent undefined index warnings

rodrigoi commented:
@kwight @Copons thanks for the feedback. I'll break it down a bit:

> One nitpick/personal preference about code organization: I find it a bit confusing that we have the main file (simple-payments.php outside the /simple-payments folder.

Actually, most of the widgets use the same pattern. A file with the widget name on the `/widgets` folder, and another folder with the resources for that widget. This is because widgets are loaded by the `widgets` module expecting those files there:

https://github.com/Automattic/jetpack/blob/24ec1ef4483f8b4d9cec3ebea3f91b63ebe00a09/modules/widgets.php#L17

we can change it, but doing so will be including another exception to the list. There are two exceptions now: `/migrate-to-core/image-widget` and `/migrate-to-core/gallery-widget`

---

> Also, since I've pinged them, I'd like to ask what are we supposed to do to bring this files into JP Fusion.
Is it enough to whitelist them and call it a day?

If I understand Fusion correctly, we'll have to make a initial manual sync. Once files are both on Jetpack and WordPress.com, then we can add them to the whitelist.

---

> I get lots of warnings and errors; maybe this is because the widget is getting registered when Simple Payments itself is not active? This site has only a personal plan (which I don't believe includes SP):

Those warnings were completely unrelated to the site's subscription. They should be fixed now that there are defaults for the `$instance` properties. 

---

>Because the widget should only be registered if the SP module is active, I'd agree that it should all be under /modules/simple-payments

If I understand correctly, the SP module is always active. The module itself never checks for an active plan or a supported feature, and the widget should follow the same pattern.
If you have a site on Premium, create a few buttons and add them to your site, and then downgrade, the buttons stay there. They are just buried using a custom post type.
The module relies on the fact that WordPress.com has the only simple interface to manage the SP Buttons... for the moment.
We should first include checks on the SP module, and add them here too.

Copons commented:
> Actually, most of the widgets use the same pattern.

@rodrigoi You're totally right, I'm not sure what I was looking at to get that wrong impression. Sorry! 🙇

Commit:
Widgets: prefixes SP widget classes with "jetpack-"

gwwar commented:
Let's open up issues for the following and have folks tackle it. Seems like these should be splittable pieces. The first item smells like a script execution issue (firing only once on the page) / or us needing to cache a DOM none. Maybe if @Copons or @vindl free up?

> - sometimes the live update removes the PayPal button.
> - if the selected Payment Button is deleted, the widget just shows the title and empty content.

Copons committed:
Simple Payments Widget: Add style overrides (#9580)

Override the media query and ensure that Simple Payments widgets are always displayed as a single column.

Copons commented:
Please rebase your local copies of this to pull in the #9580 merge!

Copons commented:
> sometimes the live update removes the PayPal button.

Taken care in #9668

Copons commented:
> if the selected Payment Button is deleted, the widget just shows the title and empty content.

Taken care in #9673

Copons committed:
Widgets: Only render the Simple Payments widget if its button exists (#9673)

In the frontend, only show the widget if the Simple Payments shortcode is parsed successfully.

In the customizer, show the widget regardless, so that it can be modified via the pencil icon.

rodrigoi committed:
Simple Payment Widget: Manage Products from the Customizer (#9699)

rodrigoi commented:
Closed in favor of https://github.com/Automattic/jetpack/pull/9782 so it can be tested with the Jetpack Beta plugin.
